# RAG 

## Evaluation!

In [1]:
import sys
import os


ruta_src = os.path.abspath(os.path.join('..', 'src'))
if ruta_src not in sys.path:
    sys.path.append(ruta_src)

from evaluation import test


tests = test.load_tests()

In [2]:
tests = test.load_tests()

In [3]:
len(tests)

21

In [4]:
example = tests[0]
print(example.question)
print(example.category)
print(example.reference_answer)
print(example.keywords)


¿Cuál es el salario actual del Ingeniero Agrónomo Alejandro Castro?
direct_fact
El salario actual de Alejandro Castro es de 42.000 €.
['Alejandro', 'Castro', '42.000 €']


In [5]:
from collections import Counter
count = Counter([t.category for t in tests])
count

Counter({'direct_fact': 6,
         'relationship': 5,
         'temporal': 4,
         'spanning': 3,
         'regulatory_compliance': 1,
         'regulatory_fact': 1,
         'traceability': 1})

In [6]:
from evaluation.eval import evaluate_retrieval, evaluate_answer

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

In [7]:
evaluate_retrieval(example)

RetrievalEval(mrr=1.0, ndcg=0.9823008098419915, keywords_found=3, total_keywords=3, keyword_coverage=100.0)

mrr=1.0 (Mean Reciprocal Rank): Es un resultado perfecto. Significa que, en promedio, la primera respuesta devuelta por el sistema fue la correcta. El MRR evalúa la posición del primer resultado relevante; un 1.0 indica que siempre estuvo en el puesto número 1.

ndcg=0.9823 (Normalized Discounted Cumulative Gain): Un puntaje de ~98.2%. Esta métrica mide la calidad general del ranking, penalizando al sistema si los resultados más relevantes aparecen abajo. Al estar tan cerca de 1.0, significa que el orden de los resultados devueltos es casi ideal.

keywords_found=3 y total_keywords=3: El sistema logró encontrar las 3 palabras clave o conceptos objetivos que se estaban buscando en la evaluación.

keyword_coverage=100.0: Directamente relacionado con lo anterior. El sistema tuvo una cobertura perfecta (100%), lo que significa que no se dejó ningún término clave importante sin recuperar.

In [8]:
eval, answer, chunks = evaluate_answer(example)

ModelResponse(id='chatcmpl-5f442daa-d2b5-440f-b64a-d8ff7da06ef4', created=1783508618, model='ollama/llama3', object='chat.completion', system_fingerprint=None, choices=[Choices(finish_reason='stop', index=0, message=Message(content='{"1. Precisi\\u00f3n": 5, "Comentario de precisi\\u00f3n": "La respuesta generada es exacta y coincide con la respuesta de referencia. El salario actual del Ingeniero Agr\\u00f3nomo Alejandro Castro se menciona como 42.000 \\u20ac en ambas respuestas, lo que indica una perfecta sincron\\u00eda entre ellas.", "2. Exhaustividad": 5, "Comentario exhaustividad": "La respuesta generada aborda de forma exhaustiva el tema solicitado. La pregunta pregunta espec\\u00edficamente por el salario actual del Ingeniero Agr\\u00f3nomo Alejandro Castro y la respuesta generada proporciona exactamente esa informaci\\u00f3n, sin omitir o agregar nada innecesario.", "3. Pertinencia": 5, "Comentario pertinencia": "La respuesta generada es directa y precisa en cuanto a lo que se 

In [9]:
eval

AnswerEval(feedback='La evaluación de la respuesta generada es la siguiente:', accuracy=5.0, completeness=4.0, relevance=5.0)

In [10]:
answer

'Según el contexto proporcionado, el salario actual del Ingeniero Agrónomo Alejandro Castro es de 42.000 €.'

In [11]:
print(eval.feedback)
print(eval.accuracy)
print(eval.completeness)
print(eval.relevance)

La evaluación de la respuesta generada es la siguiente:
5.0
4.0
5.0
